### PR TITLE
sd-bus: public API sd_bus_error_strerror

### PR DESCRIPTION
**Pull request by utezduyar**
bus_error_message is renamed to sd_bus_error_strerror and
became public.

### CODE REVIEW

File: man/sd_bus_error.xml [hunk 1]
```diff
@@ -322,6 +329,10 @@
     an error with the same <parameter>name</parameter> has been set,
     <constant>false</constant> otherwise.</para>
 
+    <para><function>sd_bus_error_strerror</function> will return the
+    error message of <parameter>e</parameter> if there is any, otherwise
+    error message of <parameter>error</parameter>.
+
```
Comment (zonque) [inline man/sd_bus_error.xml:335]: `</para>`

**Comment by dvdhrm** [comment]
Please use commit-messages to describe _why_ you do something, not _what_ you do. Anyway, your sd_bus_error_strerror() function is now declared in sd-bus.h _and_ bus-error.h. Please remove it from bus-error.h.

Furthermore, if this does _not_ make it into the release today, it must be moved into a separate section of libsystemd.sym. Just as a heads up.

**Comment by utezduyar** [comment]
Background on the PR: https://github.com/systemd/systemd/issues/224

**Comment by utezduyar** [comment]
I think it should make it to the release so people don't need to implement their own function to do what `bus_error_message` does.

**Comment by dvdhrm** [comment]
Whole patch looks good to me now. I'll leave it up to Lennart to decide..

The only issue I have with this patch is the use of `strerror()` in public APIs. It's not thread-safe and that's non-obvious for the caller of `sd_bus_error_strerror()`. We switched to `%m` everywhere except here, as it is non-trivial to do that in this case..

**Comment by poettering** [comment]
@dvdhrm makes a very good point, we are actually trying to get rid of strerror() invocations, and try to replace them with %m and strerror_r()...

But I think we can salvage things here: we can use strerror_r() instead, after all we could just modify the sd_bus_error structure and place the buffer in there, so that it is freed when the sd_bus_error object is freed.

As it turns out we even already have a call for that: sd_bus_error_set_errno(). Maybe we should base our new sd_bus_error_strerror() call on that: if the error object is already initialized we return the error message it contains. Otherwise we use sd_bus_error_set_errno() to initialized it and return the error then. That would be simple, and allows us to get rid of a major remaining user of strerror(). And especially if we declare this public API we should make this minor change.

@utezduyar could you rework the patch accordingly?

**Comment by utezduyar** [comment]
@poettering I think I have done the change you have suggested, but wouldn't it be better to do this within the APIs that take bus error like `sd_bus_call_method`. 

Ex:

```
r = sd_bus_call_method(
                                bus,
                                "org.freedesktop.machine1",
                                "/org/freedesktop/machine1",
                                "org.freedesktop.machine1.Manager",
                                "ListMachines",
                                &error,
                                &reply,
                                NULL);
if (r < 0) {
    // guaranteed error struct has the proper error message and errorno
    log_error("Could not get machines: %s", error->message);
    return r;
}
```

**Comment by poettering** [comment]
@utezduyar hmm, that makes sense actually, maybe we should ensure indeed that if an sd_bus_error structure is passed it carries useful data on error afterwards. Because right now we only pass server errors back like that, but never local errors.

Yeah, I think I agree. So I figure we should add some helper call that fills in a sd_bus_error from an error number if it is not already set, and then invoke that helper from every single function we have that returns an sd_bus_error right before exiting from it.

With such a change we should not export any new call, but just change the behaviour of the existing ones.

**Comment by utezduyar** [comment]
This PR can be closed as it is not relevant anymore. The issue https://github.com/systemd/systemd/issues/224 has been renamed to reflect the ongoing discussion.

**Pull request closed without merging**